Woo installer: use new transfer endpoints in Calypso

allilevine opened the pull request:
#### Changes proposed in this Pull Request

* Implement fetching data from the new transfer endpoints (one for Simple, one for Atomic)

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* See https://github.com/Automattic/wp-calypso/pull/58848

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

#### Error screenshot:
![VWr8UYJ8](https://user-images.githubusercontent.com/1689238/146038741-04f706f8-3f48-49b7-88ae-0902af8fa5b5.png)



Related to #58384

allilevine committed:
Set up new state

allilevine committed:
Spec out actions

allilevine committed:
Add atomic-software endpoint data layer

allilevine committed:
Split status requests

allilevine committed:
Add to atomic-software data layer

lsl commented on client/state/atomic-transfer-with-plugin/actions/index.ts:
```diff
@@ -0,0 +1,96 @@
+ * @param {string} siteId Site ID.
+ * @returns {object} An action object.
+ */
+export const initiateAtomicPluginInstall = ( softwareSet: string, siteId: string ) => ( {
```
```suggestion
export const initiateAtomicPluginInstall = ( siteId: number, softwareSet: string ) => ( {
```

It's probably best if siteId goes first for consistency with selectors, + type of siteId should be `number`.

lsl committed:
State/data-layer Refactor from #58849

- Moves datalayer to atomic/* pathing to match endpoint routing
- Mirrored the approach in state
- Adds state selectors/reducers
- todo: Replace transfer status with transfer as an object

lsl commented on client/state/data-layer/wpcom/sites/atomic/transfers/latest/index.js:
```diff
@@ -8,20 +9,20 @@ const requestTransferToAtomicStatus = ( action ) =>
-};
+const receiveResponse = ( action, response ) => [
+	setAtomicTransferStatus( action.siteId, action.softwareSet, response.status ),
+];
```
This needs reworking to store the latest transfer itself. We only need the status but I think where I've moved this to it makes more sense to make this about the whole transfer itself.

lsl commented on client/state/data-layer/wpcom/sites/atomic/transfers/index.js:
```diff
@@ -0,0 +1,46 @@
+			method: 'POST',
+			path: `/sites/${ action.siteId }/atomic/transfers/`,
+			body: {
+				software_set: action.softwareSet,
```
This parameter should be optional, we're going to need to add plugin, theme, plugin file, theme file, and later options along with it.

lsl replied:
Made this optional in the action so just needs to account for that here.

allilevine replied:
@lsl Attempted this here: https://github.com/Automattic/wp-calypso/pull/58709/commits/06e644cb1e2702e5381e0f32cf76f1fb6db1a192

lsl commented on client/state/atomic/transfers/actions.ts:
```diff
@@ -0,0 +1,55 @@
+	softwareSet,
+	meta: {
+		dataLayer: {
+			trackRequest: true,
```
We should be able to make use of this feature to track the request lifecycle in our component (e.g. isLoading checks)

lsl commented on client/state/atomic/software/reducer.js:
```diff
@@ -0,0 +1,15 @@
+export default withStorageKey(
+	'atomicSoftware',
+	keyedReducer( 'siteId', keyedReducer( 'softwareSet', software ) )
+);
```
I don't really understand these but I think I got it working in the right order. Keen to understand why we need a withStorageKey and [registerReducer](https://github.com/Automattic/wp-calypso/pull/58709/files#diff-fd71ac2664a37c14bec064251d553af36dc3daf073ddb08632505cf807076208R5) with the same key.

lsl commented on client/state/action-types.ts:
```diff
@@ -66,6 +66,15 @@ export const AUTOMATED_TRANSFER_STATUS_REQUEST_ONCE = 'AUTOMATED_TRANSFER_STATUS
+	'ATOMIC_PLUGIN_INSTALL_REQUEST_TRANSFER_STATUS';
+export const ATOMIC_PLUGIN_INSTALL_SET_TRANSFER_STATUS =
+	'ATOMIC_PLUGIN_INSTALL_SET_TRANSFER_STATUS';
+export const ATOMIC_PLUGIN_INSTALL_SET_STATUS = 'ATOMIC_PLUGIN_INSTALL_SET_STATUS';
```
I think these could use some renaming but I'm not sure what makes sense yet.

Maybe:
```
ATOMIC_SOFTWARE_INITIATE_INSTALL
ATOMIC_SOFTWARE_REQUEST_STATUS
ATOMIC_SOFTWARE_SET_STATUS
```
and
```
ATOMIC_TRANSFER_INITIATE_TRANSFER
ATOMIC_TRANSFER_REQUEST_LATEST
ATOMIC_TRANSFER_SET_LATEST
```

lsl replied:
I've updated the transfer stuff to talk about "latest transfer" instead of "status" now but software "status" is a bit murky because its the whole response, maybe we just change that to a "applied" string or an error?

lsl commented:
@allilevine had to rebase both branches to make the reset work.

I think its time I figured out how to use git worktrees.

lsl committed:
Action typing

lsl commented on client/state/atomic/software/actions.ts:
```diff
@@ -0,0 +1,78 @@
+interface AtomicSoftwareStatus {
+	blog_id: number;
+	software_set: Record< string, { path: string; state: string } >;
+	applided: boolean;
```
```suggestion
	applied: boolean;
```

lsl committed:
typo

lsl committed:
dl typing

lsl committed:
dl transfer status => latest transfer

lsl committed:
dl: error states

lsl commented on client/state/atomic/software/actions.ts:
```diff
@@ -0,0 +1,82 @@
+	blog_id: number;
+	software_set: Record< string, { path: string; state: string } >;
+	applied: boolean;
+	error?: string;
```
Don't really know how to do the error handling properly. Thoughts?

[This](https://github.com/Automattic/wp-calypso/pull/58848/files/3297b936cebf0eeabd8b8ec0102fc0bdc2324aad..26edb6b845967866cd8ee7dab3e729eb765a2111#diff-b3fec2b1927aa365b4e61090b14a29a477c73bd2b6cd22d6baa30bbc6e75beb0R35-R36) is what ends up happening in the "front end" and I think its pretty brittle. Would just creating an error selector make sense? We probably want to fix up the object format anyway (fromApi or something like that does this I think?)

lsl commented on client/state/atomic/software/actions.ts:
```diff
@@ -0,0 +1,82 @@
+ * @param {string} softwareSet Software set slug.
+ * @returns {object} An action object.
+ */
+export const requestAtomicSoftwareStatus = (
+	siteId: number,
+	softwareSet: string
+): AtomicSoftwareAction => ( {
+	type: ATOMIC_PLUGIN_INSTALL_REQUEST_STATUS,
+	siteId,
+	softwareSet,
+} );
```
```suggestion
export const requestAtomicSoftwareStatus = (
	siteId: number,
	softwareSet: string
) => ( {
	type: ATOMIC_PLUGIN_INSTALL_REQUEST_STATUS,
	siteId,
	softwareSet,
} as const );
```

https://github.com/Automattic/wp-calypso/blob/6b59de2b28bb48430ca1237b28c216b088d00662/packages/data-stores/src/auth/actions.ts#L45-L50

Thanks @sirreal

obenland commented:
@allilevine @lsl Is there anything in [client/state/atomic-transfer](https://github.com/Automattic/wp-calypso/tree/7655b6d1f326371990851da10ebb97f133fbdc17/client/state/atomic-transfer) that can be reused or built on?

lsl commented:
That looks like v1 endpoint code so I don't think so, what did you have in mind?

lsl committed:
Type fix + constant rename

lsl committed:
Transfer Status doesn't match across endpoints

lsl commented:
Happy path is stable now, testing using https://github.com/Automattic/wp-calypso/pull/58848, error handling might need an update but I think we should merge this as is to simplify / clarify whats going on in 58848.

allilevine committed:
Add README

allilevine committed:
Make software_set optional in the API call

obenland commented:
> That looks like v1 endpoint code so I don't think so, what did you have in mind?

You're right, it is :( 
It might be why `/wpcom/v2/atomic/transfer` is still unused. At one point that state was meant to be connected to it and [client/state/automated-transfer](https://github.com/Automattic/wp-calypso/tree/7655b6d1f326371990851da10ebb97f133fbdc17/client/state/automated-transfer) to be retired.

retrofox reviewed (approved):
I've tested the approach using the https://github.com/Automattic/wp-calypso/pull/58848 branch. It works as expected, keep in mind that there are still some patches to go.
Testing instruction is described in the PR.
Probably it deserves a couple of additional eyes, too.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7039878

Hi @allilevine, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `Sorry, we've hit a snag. Please contact support so we can help you out.`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7039878).
